feat: add nix flake and GitHub Actions workflow

Skyfold opened the pull request:
This is about creating a Nix flake that: 

1. Can be used by hyperbole's Nix flake to build with the same dependencies (the overlay provides this)
2. Provide an easy way to test out the example project with nix `nix develop .#example`
3. Provide CI integration for the library
4. Provide a development shell for web-view
5. Also adds the test/resources to the cabal file so Hackage gets them.

Since the golden tests do not pass, the `nix develop .#example` comamnd will fail (so does the CI). I decided to put those fixes in a separate pull request since they are code changes.

More detail:

* Add Nix flake configuration for building and development
  - Configure build inputs and overlays
  - Set up development shells for main package and example

* Add GitHub Actions workflow for CI
  - Build packages and devShells on Ubuntu and macOS
  - Test overlay application
  - Use DeterminateSystems' Nix installer and cache actions

* Update documentation
  - Add Local Development section with Nix instructions
  - Include example usage of flake overlay
  - Document dev shell and example project setup

* Fix cabal file
  - Add missing test resource files to extra-source-files

Skyfold commented:
The test fixes are minor rendering changes to match the golden test. I went ahead and added the fix here.

seanhess commented:
Looks good!

I recently swapped out the nix github action in github, because it was failing and I didn't know how to maintain it. I went with a simple docker-based action. It's slow, but easy to maintain, and consistent. https://github.com/seanhess/hyperbole/blob/main/.github/workflows/haskell.yaml

Are you willing to keep the CI running for this? Or should we switch to the above docker version?

Skyfold commented:
I'm happy to keep the CI running, since I need the flake myself. I can also write docs on how to maintain the flake.nix.

Skyfold commented:
I commented out the hpack pre-commit check because it fails due to using a different version of hpack. The newest hpack version in nixpkgs is 0.36.1 while you are using 0.37.0. I'll put it back when our hpack versions are the same.

Skyfold commented:
I needed to pull out the build-tool-depends in the cabal file:

```
  build-tool-depends:
      skeletest:skeletest-preprocessor
```

I could not get cabal to realise it was already in the PATH

Skyfold commented:
I finished the rebase to clean up the commits, it is ready to merge.

Once it is merged I'll put up the pull request for hyperbole with its flake.

Thanks!

Skyfold commented:
I noticed a `.pre-commit-config.yaml` slipped in, I'll add an extra commit to pull that out, one sec.

Skyfold commented:
fixed

seanhess commented on package.yaml:
```diff
@@ -19,6 +19,12 @@ extra-doc-files:
 
 extra-source-files:
   - embed/preflight.css
+  - test/resources/basic.txt
```
Doesn't this add these text files to hackage? Why do we need them in extra-source-files when they are only used in tests?

Skyfold replied:
If you pull down the source from Hackage, then the tests fail without those files. Since I pull the web-view source from github with the flake, I don't need those files specifed in the cabal file.

Skyfold replied:
let me know if you still want to remove them from extra-source-files.

seanhess replied:
I'd rather they be removed from the extra-source-files if it won't mess up your approach here.

seanhess commented on src/Web/View/Render.hs:
```diff
@@ -79,12 +79,12 @@ renderText' c vw =
   addCss [] cnt = cnt
   addCss css cnt = do
-    renderLines (renderContent 2 $ styleElement css) <> "\n\n" <> cnt
+    renderLines (renderContent 0 $ styleElement css) <> "\n\n" <> cnt
```
This is going to need merging with the new main, which has more explicit newlines

Skyfold replied:
Can do, your fix makes the intention clear.

seanhess reviewed (changes requested):
nice work!

Skyfold commented:
This is ready to merge (pending extra-source-files question). Note, the first time it builds on CI around 30 minutes. Unless we change dependencies, the cache brings that down to around 3 minutes for linux and 13 minutes for macos. Not sure why the macos is so much slower.

Skyfold commented:
In hyperbole you have:

```
tested-with:
  - GHC == 9.8.2
  - GHC == 9.6.6
```

Do you want to have the same for web-view?

I can setup CI to test both GHC versions.

Skyfold commented:
I went ahead and set it up for ghc982 and ghc966. This is good to merge.

seanhess commented:
Great! Let's get those extra-source-files removed then and merge. Thanks!

Skyfold commented:
Removed the golden test files from `extra-source-files`.

Adding ghc982 did slow down the CI quite a bit more than I expected. I removed testing the development shells to speed it up. I think eventually we will want to setup [cachix](https://github.com/cachix/cachix-action) for the CI and make it fast for users to try out the library (so they don't have to wait to build it themselves). Cachix is free for open source projects, but it does require creating an accound and adding their auth token to github workflow secrets.

Once I see the tests are passing for macos I'll say its good to merge.

Skyfold commented:
This is ready to merge.

I got the CI times down to 10 minutes after the development shell tests where removed. I also removed testing on macos, there is a linking issue in nix for ghc982 only for the macos platform.

Skyfold commented:
Minor update, the `nix check` wasn't testing the example code.
